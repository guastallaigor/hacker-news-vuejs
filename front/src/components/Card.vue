<template functional>
  <div class="card" :style="{ backgroundImage: `url(${props.url})` }">
    <h1 class="text-shadow">{{ props.title }}</h1>
  </div>
</template>

<style lang="scss">
.card {
  text-align: center;
  margin-right: 2em;
  width: 400px;
  cursor: pointer;
  background-size: cover !important;
  border-radius: 10px;
  background-position: center !important;
  color: white;
  height: 200px;
  padding: 1em 2em;
  position: relative;
  box-shadow: 0 0 5px rgba(0,0,0,.3),
      0 0 10px rgba(0,0,0,.1);
  will-change: box-shadow, top;
  transition: top .25s ease-out;
  top: 0;

  > h1 {
    font-size: 3.9em;
  }

  &:hover {
    box-shadow: 0 0 15px rgba(0,0,0,.4),
      0 0 25px rgba(0,0,0,.2);
    top: -5px;
  }
}

.text-shadow {
  text-shadow: 1px 1px rgba(0, 0, 0, 0.04), 2px 2px rgba(0, 0, 0, 0.04),
    3px 3px rgba(0, 0, 0, 0.04), 4px 4px rgba(0, 0, 0, 0.04),
    0.125rem 0.125rem rgba(0, 0, 0, 0.04), 6px 6px rgba(0, 0, 0, 0.04),
    7px 7px rgba(0, 0, 0, 0.04), 8px 8px rgba(0, 0, 0, 0.04), 9px 9px rgba(0, 0, 0, 0.04),
    0.3125rem 0.3125rem rgba(0, 0, 0, 0.04), 11px 11px rgba(0, 0, 0, 0.04),
    12px 12px rgba(0, 0, 0, 0.04), 13px 13px rgba(0, 0, 0, 0.04),
    14px 14px rgba(0, 0, 0, 0.04), 0.625rem 0.625rem rgba(0, 0, 0, 0.04),
    16px 16px rgba(0, 0, 0, 0.04), 17px 17px rgba(0, 0, 0, 0.04),
    18px 18px rgba(0, 0, 0, 0.04), 19px 19px rgba(0, 0, 0, 0.04),
    1.25rem 1.25rem rgba(0, 0, 0, 0.04);
}
</style>
