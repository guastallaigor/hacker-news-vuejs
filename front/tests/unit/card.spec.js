import { shallowMount, mount } from '@vue/test-utils'
import Card from '@/components/Card'

describe('When I create the Card component', () => {
  const card = {
    title: 'Card title',
    url: 'https://www.gravatar.com/avatar?s=15'
  }
  const createCardWrapper = (propsData = {}, slot, real = false) => {
    const obj = {
      propsData,
      slots: {
        default: slot || 'Card'
      }
    }

    return real ? mount(Card, { ...obj }) : shallowMount(Card, { ...obj })
  }

  it(`should have a title if the card prop is added with the title attribute,
      and has a showTitle prop true, and have a text-shadow class`, () => {
    const wrapper = createCardWrapper({ card, showTitle: true })
    const title = wrapper.find('.card>h1')
    expect(wrapper.exists()).toBe(true)
    expect(title.exists()).toBe(true)
    expect(title.classes().length).toBe(1)
    expect(title.classes()).toContainEqual('text-shadow')
  })

  it(`should have a background-image if the card prop is added with the url attribute`, () => {
    const wrapper = createCardWrapper({ card })
    const cardDiv = wrapper.find('.card')
    expect(wrapper.exists()).toBe(true)
    expect(cardDiv.exists()).toBe(true)
    expect(cardDiv.attributes().style).toBe('background-image: url(https://www.gravatar.com/avatar?s=15);')
  })

  it(`should have a less-padding class if the lessPadding prop is added and set to true`, () => {
    const wrapper = createCardWrapper({ card, lessPadding: true })
    const cardDiv = wrapper.find('.card')
    expect(wrapper.exists()).toBe(true)
    expect(cardDiv.exists()).toBe(true)
    expect(cardDiv.classes().length).toBe(2)
    expect(cardDiv.classes()).toContainEqual('less-padding')
  })

  it(`should have a not-clickable class if the notClickable prop is added and set to true`, () => {
    const wrapper = createCardWrapper({ card, notClickable: true })
    const cardDiv = wrapper.find('.card')
    expect(wrapper.exists()).toBe(true)
    expect(cardDiv.exists()).toBe(true)
    expect(cardDiv.classes().length).toBe(2)
    expect(cardDiv.classes()).toContainEqual('not-clickable')
  })

  it(`should have a not-clickable class if the notClickable prop is added and set to true`, () => {
    const wrapper = createCardWrapper({ card, notClickable: true })
    const cardDiv = wrapper.find('.card')
    expect(wrapper.exists()).toBe(true)
    expect(cardDiv.exists()).toBe(true)
    expect(cardDiv.classes().length).toBe(2)
    expect(cardDiv.classes()).toContainEqual('not-clickable')
  })

  it('should print html inside slot', () => {
    const html = '<p id="find-this-id">This is a card slot</p>'
    const wrapper = createCardWrapper({}, html)
    const slot = wrapper.find('p#find-this-id')

    expect(wrapper.exists()).toBe(true)
    expect(slot.exists()).toBe(true)
    expect(slot.html()).toBe(html)
  })

  it('should emit a click event when clicked on a card', () => {
    const wrapper = createCardWrapper({ card }, null, true)

    const cardDiv = wrapper.find('.card')
    cardDiv.trigger('click')
    cardDiv.trigger('click')
    cardDiv.trigger('click')

    expect(wrapper.exists()).toBe(true)
    expect(cardDiv.exists()).toBe(true)
    setTimeout(() => {
      expect(wrapper.emitted('click')).toBeTruthy()
      expect(wrapper.emitted('click').length).toBe(3)
    })
  })

  it('should match snapshot', () => {
    const wrapper = createCardWrapper({ card, showTitle: true, lessPadding: true })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
