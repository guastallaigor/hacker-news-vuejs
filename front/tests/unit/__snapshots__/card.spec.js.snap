// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When I create the Card component should match snapshot 1`] = `
<div class="card less-padding" style="background-image: url(https://www.gravatar.com/avatar?s=15);">
  <h1 class="text-shadow">Card title</h1> Card
</div>
`;
